engine/table/src/main/java/io/deephaven/engine/table/impl/locations/local/URIStreamKeyValuePartitionLayout.java

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * Extracts a Key-Value partitioned layout from a stream of URIs. This class is inspired from
+ * {@link KeyValuePartitionLayout}, but for URIs instead of files.
+ */
+public class URIStreamKeyValuePartitionLayout<TLK extends TableLocationKey> {
+
+    protected final URI tableRootDirectory;
+    private final Supplier<KeyValuePartitionLayout.LocationTableBuilder> locationTableBuilderFactory;
+    private final BiFunction<URI, Map<String, Comparable<?>>, TLK> keyFactory;
+    private final int maxPartitioningLevels;
+    private final String fileSeparator;
+
+    public URIStreamKeyValuePartitionLayout(
+            @NotNull final URI tableRootDirectory,
+            @NotNull final Supplier<KeyValuePartitionLayout.LocationTableBuilder> locationTableBuilderFactory,
+            @NotNull final BiFunction<URI, Map<String, Comparable<?>>, TLK> keyFactory,
+            final int maxPartitioningLevels,
+            @NotNull final String fileSeparator) {
+        this.tableRootDirectory = tableRootDirectory;
+        this.locationTableBuilderFactory = locationTableBuilderFactory;
+        this.keyFactory = keyFactory;
+        this.maxPartitioningLevels = Require.geqZero(maxPartitioningLevels, "maxPartitioningLevels");
+        this.fileSeparator = fileSeparator;
+    }
+
+    public String toString() {
+        return URIStreamKeyValuePartitionLayout.class.getSimpleName() + '[' + tableRootDirectory + ']';
+    }
+
+    /**
+     * Find the keys in the given URI stream and notify the observer.
+     */
+    protected final void findKeys(@NotNull final Stream<URI> uriStream,
+            @NotNull final Consumer<TLK> locationKeyObserver) {
+        final KeyValuePartitionLayout.LocationTableBuilder locationTableBuilder = locationTableBuilderFactory.get();
+        final Deque<URI> targetURIs = new ArrayDeque<>();
+        final Set<String> takenNames = new HashSet<>();
+        final List<String> partitionKeys = new ArrayList<>();
+        final boolean[] registered = {false}; // Hack to make the variable final
+        // TODO Should I use something that orders the operations instead of forEach, like convert it to an array first?
+        uriStream.forEach(uri -> {
+            final Collection<String> partitionValues = new ArrayList<>();
+            final String fileRelativePath = uri.getPath().substring(tableRootDirectory.getPath().length());
+            getPartitions(fileRelativePath, partitionKeys, partitionValues, takenNames, registered[0]);
+            if (!registered[0]) {
+                locationTableBuilder.registerPartitionKeys(partitionKeys);
+                registered[0] = true;
+            }
+            locationTableBuilder.acceptLocation(partitionValues);
+            targetURIs.add(uri);
+        });
+        final Table locationTable = locationTableBuilder.build();
+        buildLocationKeys(locationTable, targetURIs, locationKeyObserver, keyFactory);
```

Comment:
I don't really understand the full pattern here; Ryan will want to make sure we are building this efficiently.